### PR TITLE
[21137] RTPS writer APIs refactor

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

This is a refactor on the RTPS writer APIs, so only the ones intended for the user (i.e. public and exported) are present.
The refactor includes adding an intermediate `BaseWriter` class with the internal APIs.

Related PRs:
* https://github.com/eProsima/ShapesDemo/pull/159
* https://github.com/eProsima/Fast-DDS-python/pull/157

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- _N/A_ Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- _N/A_ Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- ❌ Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- ❌ Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 
    - Related documentation PR: eProsima/Fast-DDS-docs#847
- _N/A_ Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- _N/A_: If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/publisher/DataWriterImpl.hpp
Comment (EduPonz) [inline src/cpp/fastdds/publisher/DataWriterImpl.hpp:1]: I think in this file there are some namespacing uses that can be removed/shortened, such as `fastdds::rtps::MatchingInfo`, which could be `rtps::MatchingInfo`

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/writer/BaseWriter.hpp [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+        static_cast<void>(seq_num);
+        return false;
+    }
+
+    FASTDDS_EXPORTED_API bool wait_for_all_acked(
+            const Duration_t& max_wait) override
+    {
+        static_cast<void>(max_wait);
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/writer/BaseWriter.hpp:89]: They are already (and marked override)

File: src/cpp/rtps/writer/BaseWriter.hpp [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+
+    FASTDDS_EXPORTED_API bool is_async() const override
+    {
+        return is_async_;
+    }
+
+    FASTDDS_EXPORTED_API bool get_disable_positive_acks() const override
+    {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/writer/BaseWriter.hpp:111]: They are already (and marked override)

File: src/cpp/rtps/writer/BaseWriter.hpp [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+     */
+    FASTDDS_EXPORTED_API void set_enabled_statistics_writers_mask(
+            uint32_t enabled_writers) override;
+
+#endif // FASTDDS_STATISTICS
+
+    WriterHistory* get_history() const
+    {
+        return history_;
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/writer/BaseWriter.hpp:144]: I kept the non-virtual one-liner getter methods in the header and marked them as inline.

File: src/cpp/rtps/writer/BaseWriter.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/writer/BaseWriter.hpp]: Sorry, I meant pure virtual. We agreed on removing the impl of those methods that have different behaviour for stateless and stateful writers from BaseWriter

File: src/cpp/rtps/writer/BaseWriter.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/writer/BaseWriter.hpp]: https://github.com/eProsima/Fast-DDS/pull/5028#discussion_r1670473417

File: src/cpp/rtps/writer/BaseWriter.hpp [hunk 4]
```diff
@@ -67,68 +67,31 @@ class BaseWriter
...
+    //vvvvvvvvvvvvvvvvvvvvv [Exported API] vvvvvvvvvvvvvvvvvvvvv
 
-    FASTDDS_EXPORTED_API bool is_acked_by_all(
-            const SequenceNumber_t& seq_num) const override;
+    WriterListener* get_listener() const final;
 
-    FASTDDS_EXPORTED_API bool wait_for_all_acked(
-            const Duration_t& max_wait) override;
+    bool set_listener(
+            WriterListener* listener) final;
```
Comment (EduPonz) [inline src/cpp/rtps/writer/BaseWriter.hpp:75]: This way, if the base function ceases to exist or it's made non-virtual we won't compile. Please apply to all inherited member functions. Applies to other files
```suggestion
            WriterListener* listener) override final;
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/writer/BaseWriter.hpp (hunk 4 above)
Comment (EduPonz) [reply src/cpp/rtps/writer/BaseWriter.hpp:75]: We decided that since the functions are not marked virtual on the derived class we don't need it.

**Comment by EduPonz** [comment]
CI passed for f421bf1

**Review by EduPonz — APPROVED** [review]

**Pull request merged**